### ISSUE TITLE
3.0 when will it be released

### ISSUE THREAD

**Issue by taodaqiao** [Bareflank/hypervisor#945]
3.0 when will it be released

**Comment by rianquinn** [comment]
@taodaqiao Thanks a ton for the interest in our project. 3.0 should be released soon. Most of the code is complete, but we are working on the following before releasing an official pre-release:
- Unit Tests for each component. Not looking for get to 100%, but just enough to prove that the code layout, APIs, etc. are all unit testable. 
- ARM support. We have the loader complete, but the Microkernel still needs some work done. This should not take long as most of the hard work is complete (still have one researchy thing to figure out with how UEFI handles moving to EL1). 
- Failure Logic. There are still a couple of syscall handlers that do not have a functional failure handler, including the VPS, mem and intrinsic ops. 
- Final cleanup. 

Once this is done, I will be creating an official pre-release. The final release will of course come once all of the unit tests, integration tests and system tests are complete, and everyone seems happy with what has been complete. Right now, I am in the middle of a massive patch that will address a number of issues, adds ARM support, etc. I was hoping to have it pushed this weekend, but have run into a number of issues that the initial unit tests have identified that will change how the syscall API (not the  ABI) is handled. Hopefully I will have the push complete this week so that you can play with what we have.

**Comment by danyhm** [comment]
I too am waiting impatiently for the 3.0 
The Docs should definitely be improved.

**Comment by xiaolongwu01** [comment]
Does the Microkernel in Bareflank Hypervisor support multi-core?
What functions does Microkernel currently support?

**Comment by rianquinn** [comment]
@xiaolongwu01 Yes, multicore is supported. The APIs that Bareflank supports right now can be seen here, with an update coming soon that adds some additional memory ops:

https://github.com/Bareflank/hypervisor/blob/master/docs/Microkernel%20Syscall%20Specification.md

**Comment by rianquinn** [comment]
@danyhm Any suggestions on how to improve the docs? As for 3.0, we are definitely trying our best, and it is getting close. Close enough that you should be able to try it out now. Nothing is stable yet, but again, it is getting close.

**Comment by danyhm** [comment]
@rianquinn Thanks for asking. well, you mentioned that every SDK function will be documented by Doxygen and honestly I don't think that's enough because the topic of virtualization is far more complex than that. I think step-by-step tutorials will best explain how to work with the SDK. also a general visualization of different parts of the code will help us find what each part does better. 
I first learned about Bareflank in this blog:
https://rayanfam.com/topics/hypervisor-from-scratch-part-1/

Since Bareflank is an SDK, it's not needed to go that deep into what's going on for every step but I think more focus on the application level explanation is needed.

for example, in my own project, I need to analyze a binary that uses the CPUID instruction to detect the CPU information. I need Bareflank to give me control of that instruction so I can manipulate the CPUID results. of course, I don't expect the docs to explain this specific scenario but I want to know:

1- how can I start a simple project?
2- Is this even possible? what can and can't be done with Bareflank (example cases)
3- how to proceed to use the SDK to achieve some specific task?

I hope i made my point as clear as i could.

**Comment by rianquinn** [comment]
@danyhm Absolutely, and I really appreciate the feedback. 
1) https://github.com/Bareflank/hypervisor#writing-your-own-extensions
2) What can/cannot be done with Bareflank is pretty much limited to what can/cannot be done with Virtualization, so it is more up to you. We do plan on providing as many examples as possible. We don't have many yet (working on that), but the examples folder will have some good starting points that people can pull from (and in your case, CPUID should be one of them). 
3). Can you provide more clarification on this?  Maybe an example of what you are looking for here. 

Also, for your specific case, you will need to modify this code:
https://github.com/Bareflank/hypervisor/blob/master/example/default/src/x64/intel/vmexit_t.hpp#L302

This happens to be the case for you because all extensions need to handle CPUID. Probably the best advice I can give is really learn the examples and how they work. I have a TON of comments in the them, so if more comments are needed or something doesn't make sense, please let me know and I can add more detail.

**Comment by danyhm** [comment]
@rianquinn thanks.
For #3 , I mean how to do things which are possible to be done with the SDK. In my case you mentioned the code i need to edit, well what about other parts?

For example if you lookup Telegram Bots API you'll see that for every possible API there's an explaination along with an example for every series of command. Although for Bareflank this may not be possible since like you said there's a ton of stuff possible that are limited by the virtualization as well.

Maybe we should defer this part because it might get too big.

**Comment by rianquinn** [comment]
@danyhm I think it would likely be a good idea to at least document an example of how to use the APIs that are provided by the syscall layer. This way, you can at least see how to call each syscall. We definitely cannot provided every possible use of the API as that is pretty big, but I agree that at least one example of each syscall on how to call it would be a good idea.

**Comment by danyhm** [comment]
@rianquinn that would be awesome! btw I checked the default examples and i couldn't find how to virtualize an existing .exe windows file using the SDK.
base on your comment in https://github.com/Bareflank/MicroV/issues/22 I thought that it's possible just to virtualize a single .exe windows binary but as I've studied since then I can't seem to find how that's possible.
can you please shed light on this?

**Comment by rianquinn** [comment]
https://github.com/Bareflank/hypervisor/releases/tag/v3.0.0
